In this project, we examine a page visits funnel for an online T-shirt company to examine how many people complete a 4-step process:
<ol>1.  A user visits CoolTShirts.com</ol>
<ol>2.  A user adds a t-shirt to their cart.</ol>
<ol>3.  A user clicks "checkout".</ol>
<ol>4.  A user completes their purchase.</ol>

This project will be completed using the pandas library within python and will involve 4 different csv sheets. Let's get started.
<br>
<br>
First, we import pandas.

In [1]:
# Import pandas
import pandas as pd

Next, we import our 4 csv sheets into pandas as dataframes. We'll also use the parse_dates functionality to make sure our data is rendered properly:

In [2]:
# Import the 4 csv files as pandas dataframes.
visits = pd.read_csv('visits.csv',
                     parse_dates=[1])
cart = pd.read_csv('cart.csv',
                   parse_dates=[1])
                   
checkout = pd.read_csv('checkout.csv',
                       parse_dates=[1])
purchase = pd.read_csv('purchase.csv',
                       parse_dates=[1])

It's a good idea to check if there are any duplicate rows in any of the DataFrames. We can do this efficiently like this:

In [3]:
dataframes = [visits, cart, checkout, purchase]

for df in dataframes:
    print(len(df)==len(df.drop_duplicates()))

True
True
True
False


Let's fix this:

In [5]:
purchase.drop_duplicates(inplace=True)

purchase.reset_index(inplace=True)

Next, we'll print the first 5 lines of the DataFrames to inspect them:

In [6]:
print(visits.head())
print(cart.head())
print(checkout.head())
print(purchase.head())

                                user_id          visit_time
0  943647ef-3682-4750-a2e1-918ba6f16188 2017-04-07 15:14:00
1  0c3a3dd0-fb64-4eac-bf84-ba069ce409f2 2017-01-26 14:24:00
2  6e0b2d60-4027-4d9a-babd-0e7d40859fb1 2017-08-20 08:23:00
3  6879527e-c5a6-4d14-b2da-50b85212b0ab 2017-11-04 18:15:00
4  a84327ff-5daa-4ba1-b789-d5b4caf81e96 2017-02-27 11:25:00
                                user_id           cart_time
0  2be90e7c-9cca-44e0-bcc5-124b945ff168 2017-11-07 20:45:00
1  4397f73f-1da3-4ab3-91af-762792e25973 2017-05-27 01:35:00
2  a9db3d4b-0a0a-4398-a55a-ebb2c7adf663 2017-03-04 10:38:00
3  b594862a-36c5-47d5-b818-6e9512b939b3 2017-09-27 08:22:00
4  a68a16e2-94f0-4ce8-8ce3-784af0bbb974 2017-07-26 15:48:00
                                user_id       checkout_time
0  d33bdc47-4afa-45bc-b4e4-dbe948e34c0d 2017-06-25 09:29:00
1  4ac186f0-9954-4fea-8a27-c081e428e34e 2017-04-07 20:11:00
2  3c9c78a7-124a-4b77-8d2e-e1926e011e7d 2017-07-13 11:38:00
3  89fe330a-8966-4756-8f7c-3bdbcd47279a 

We notice that each DataFrame has a similar structure: a user_id column and a second column indicating a certain action was taken by the customer (when they first logged on, when they added something to their cart, when they checked out, and when they completed their purchase).
<br>
<br>
To see how many people who visited the website added an item to their cart, we can begin by doing a left merge on the first two DataFrames, visits and carts. A left merge will make sure all data from the lefthand DataFrame (in this case, visits) will be included in the merged DataFrame, while NaN values will be filled in for anyone who did not add something to their cart:

In [7]:
# Create a left-merge between the visits and carts DataFrames
visit_cart_merge = pd.merge(visits, cart, how='left')

print(visit_cart_merge.head())

                                user_id          visit_time  \
0  943647ef-3682-4750-a2e1-918ba6f16188 2017-04-07 15:14:00   
1  0c3a3dd0-fb64-4eac-bf84-ba069ce409f2 2017-01-26 14:24:00   
2  6e0b2d60-4027-4d9a-babd-0e7d40859fb1 2017-08-20 08:23:00   
3  6879527e-c5a6-4d14-b2da-50b85212b0ab 2017-11-04 18:15:00   
4  a84327ff-5daa-4ba1-b789-d5b4caf81e96 2017-02-27 11:25:00   

            cart_time  
0                 NaT  
1 2017-01-26 14:44:00  
2 2017-08-20 08:31:00  
3                 NaT  
4                 NaT  


To get the length of each dataframe, we can use the len() method:

In [8]:
# Print the length of visits, carts, and their merge
print(len(visits), len(cart), len(visit_cart_merge))

2000 348 2000


As we can see, the left merge we performed has the same length (2000 rows) as the visits DataFrame, while the carts DataFrame is much shorter. This means that of the 2000 people who visited the website, 1652 of them did not add anything to their cart. We can also see this by checking the number of NaN values in the merged_data frame in the 'cart_time' category.

In [9]:
# Print the number of rows of the merged DataFrame that has a null value for the cart_time column
print(len(visit_cart_merge[visit_cart_merge['cart_time'].isnull()]))

1652


The values match as expected, the percent of visitors who added something to their cart is:

In [10]:
# Calculate the percent of visitors who added something to their cart
print(str((348/2000)*100) + "%")

17.4%


Only 17.4% of 2000 visitors added an item to their cart or 82.6% people of visitors didn't add anything to their cart.
<br>
<br>
Now let's see how many people proceeded to checkout after adding something to their cart. We'll perform another left merge but this time with the cart and checkout DataFrames:

In [11]:
# Perform a left merge on the cart and checkout DataFrames
cart_checkout_merge = pd.merge(cart, checkout, how='left')

# Print the length of the cart + checkout merge and the number of rows where the checkout_time column is null
print(len(cart_checkout_merge), len(cart_checkout_merge[cart_checkout_merge['checkout_time'].isnull()]))

348 122


This merge has 348 rows (as expected, sicne we performed a left merge with cart being the left DataFrame) and 122 rows with a null value for the checkout time. As a percent this is:

In [12]:
# Calculate the percentage of people who add something to their cart continue to checkout
print(str((122/348)*100) + "%")

35.05747126436782%


~35% of people who added something to their cart did not proceed to check out. In other words, 65% of people who added something to their cart <em>did</em> check out.
<br>
<br>
Now, let's see how many of the people who proceed to checkout actually complete their purchase. We'll perform a left merge on the checkout and purchase dataframes:

In [13]:
# Perform a left merge on the checkout and purchase DataFrames
checkout_purchase_merge = pd.merge(checkout, purchase, how='left')

# Print the length of the c + checkout merge and the number of rows where the checkout_time column is null
print(len(checkout_purchase_merge), len(checkout_purchase_merge[checkout_purchase_merge['purchase_time'].isnull()]))

330 82


In [15]:
# Calculate the percentage of people who proceed to checkout complete the purchase
print(str((82/330)*100) + "%")

24.848484848484848%


Of the people who proceed to checkout, 24.85% did not complete their purchase or 75.15% of those who proceeded to checkout completed their purchase.
<br>
<br>
Clearly, the weakest part of the funnel is getting someone who visits the website to add something to their cart.

*The weakest part of the funnel is clearly getting a person who visited the site to add a tshirt to their cart. Once they've added a t-shirt to their cart it is fairly likely they end up purchasing it. A suggestion could be to make the add-to-cart button more prominent on the front page.*


For those who do complete their purchase, let's see how long they take from the time they visit the site to completing the purchase. To do this, we'll need to merge the all the data_frames:

In [16]:
# Perform a left merge on the first 3 DataFrames
visit_cart_checkout_merge = pd.merge(visit_cart_merge, checkout, how='left')

# Perform a left merge on all the DataFrames
all_data = pd.merge(visit_cart_checkout_merge, purchase, how='left')

To calculate the average time from arriving on the site to completing the purchase, let's first create a new column:

In [22]:
# Create a new column showing the amount of time elapsed between when someone visits the site and when they complete their purchase
all_data['time_to_complete_purchase'] = all_data['purchase_time'] - all_data['visit_time']

# Display the first 5 rows of all_data
print(all_data.head())

                                user_id          visit_time  \
0  943647ef-3682-4750-a2e1-918ba6f16188 2017-04-07 15:14:00   
1  0c3a3dd0-fb64-4eac-bf84-ba069ce409f2 2017-01-26 14:24:00   
2  6e0b2d60-4027-4d9a-babd-0e7d40859fb1 2017-08-20 08:23:00   
3  6879527e-c5a6-4d14-b2da-50b85212b0ab 2017-11-04 18:15:00   
4  a84327ff-5daa-4ba1-b789-d5b4caf81e96 2017-02-27 11:25:00   

            cart_time       checkout_time  index       purchase_time  \
0                 NaT                 NaT    NaN                 NaT   
1 2017-01-26 14:44:00 2017-01-26 14:54:00  121.0 2017-01-26 15:08:00   
2 2017-08-20 08:31:00                 NaT    NaN                 NaT   
3                 NaT                 NaT    NaN                 NaT   
4                 NaT                 NaT    NaN                 NaT   

  time_to_complete_purchase  
0                       NaT  
1           0 days 00:44:00  
2                       NaT  
3                       NaT  
4                       NaT  


Now let's get the summary statistics on this last column:

In [23]:
all_data.time_to_complete_purchase.describe()

count                          248
mean     0 days 00:43:07.741935483
std      0 days 00:14:39.083842854
min                0 days 00:09:00
25%                0 days 00:32:00
50%                0 days 00:44:30
75%                0 days 00:54:00
max                0 days 01:17:00
Name: time_to_complete_purchase, dtype: object

We can see that the average time to complete a purchase is about 43 minutes. 75% of customers complete their purchase in just under an hour. This would seem to indicate that people are browsing on the website before purchasing rather than knowing immediately what they want upon visiting.